From 9d9ef9b0b9a3f85156aa298a25aa0a330fb492a6 Mon Sep 17 00:00:00 2001
From: "Subhransu S. Prusty" <subhransu.s.prusty@intel.com>
Date: Mon, 21 Sep 2015 18:13:00 +0530
Subject: [PATCH 1669/2508] Set the dma data to NULL in hw_free

commit 25d8226a3626536fe27332a5bd07e515842df60c from
https://github.com/01org/linux-apollolake-i

Also hw_free is called two times, 1) from userspace and 2) from substream
pcm release, so add a check for link dev to avoid any incorrect memory
access.

Signed-off-by: Subhransu S. Prusty <subhransu.s.prusty@intel.com>
---
 sound/soc/intel/skylake/skl-pcm.c |    4 ++++
 1 files changed, 4 insertions(+), 0 deletions(-)

diff --git a/sound/soc/intel/skylake/skl-pcm.c b/sound/soc/intel/skylake/skl-pcm.c
index aed9284..8fb723a 100644
--- a/sound/soc/intel/skylake/skl-pcm.c
+++ b/sound/soc/intel/skylake/skl-pcm.c
@@ -426,6 +426,9 @@ static int skl_link_hw_free(struct snd_pcm_substream *substream,
 				snd_soc_dai_get_dma_data(dai, substream);
 	struct hdac_ext_link *link;
 
+	if (!link_dev)
+		return 0;
+
 	dev_dbg(dai->dev, "%s: %s\n", __func__, dai->name);
 
 	link_dev->link_prepared = 0;
@@ -434,6 +437,7 @@ static int skl_link_hw_free(struct snd_pcm_substream *substream,
 	if (!link)
 		return -EINVAL;
 
+	snd_soc_dai_set_dma_data(dai, substream, NULL);
 	snd_hdac_ext_link_clear_stream_id(link, hdac_stream(link_dev)->stream_tag);
 	snd_hdac_ext_stream_release(link_dev, HDAC_EXT_STREAM_TYPE_LINK);
 	return 0;
-- 
1.7.5.4

